LoggerMixin instead of logger.connect()

happz opened the pull request:
Using a mixin class we can get rid of "logging type stubs" because,
suddenly, all logging methods are inherited from the mixin class \o/

It required few tweaks to get there - our ContextAdapter gained its own
set of logging methods, overloading those inherited from its parent
class, because 1) logging's idea about signature of these methods didn't
play well with exception info returned by sys.exc_info(), and 2) we like
our sentry=True parameter to submit warnings.

So, LoggerMixin is used to bring methods to classes. Sentry's "submit
warning" become "submit message" because it's no longer just a warning,
sentry parameter is supported everywhere - this and explicitly stating
that ContextAdapter is our main logging vehicle let's us get rid of
patching logging with our versions of warn() and verbose(). All our code
is using ContextAdapter and its children, attaching their methods to
classes via LoggerMixin, and Logger instances used by 3rd party
libraries are simply redirected to our handlers.

Things are much simler now, so happy, such wow.

happz committed:
LoggerMixin instead of logger.connect()

Using a mixin class we can get rid of "logging type stubs" because,
suddenly, all logging methods are inherited from the mixin class \o/

It required few tweaks to get there - our ContextAdapter gained its own
set of logging methods, overloading those inherited from its parent
class, because 1) logging's idea about signature of these methods didn't
play well with exception info returned by sys.exc_info(), and 2) we like
our sentry=True parameter to submit warnings.

So, LoggerMixin is used to bring methods to classes. Sentry's "submit
warning" become "submit message" because it's no longer just a warning,
sentry parameter is supported everywhere - this and explicitly stating
that ContextAdapter is our main logging vehicle let's us get rid of
patching logging with our versions of warn() and verbose(). All our code
is using ContextAdapter and its children, attaching their methods to
classes via LoggerMixin, and Logger instances used by 3rd party
libraries are simply redirected to our handlers.

Things are much simler now, so happy, such wow.

thrix commented on gluetool/log.py:
```diff
@@ -580,62 +483,165 @@ def removeHandler(self, *args, **kwargs):
+        self.log(logging.DEBUG, msg, exc_info=exc_info, extra=extra, sentry=sentry)
+
+    # pylint: disable=arguments-differ
+    def info(self, msg, exc_info=None, extra=None, sentry=False):   # type: ignore  # Signature of "debug" incompatible with supertype "LoggerAdapter"
```
signature of info ...

happz replied:
Fixed - explanation is now shared.

thrix commented on gluetool/log.py:
```diff
@@ -580,62 +483,165 @@ def removeHandler(self, *args, **kwargs):
+        self.log(logging.INFO, msg, exc_info=exc_info, extra=extra, sentry=sentry)
+
+    # pylint: disable=arguments-differ
+    def warning(self, msg, exc_info=None, extra=None, sentry=False):   # type: ignore  # Signature of "debug" incompatible with supertype "LoggerAdapter"
```
same here

happz replied:
Fixed - explanation is now shared.

thrix commented on gluetool/log.py:
```diff
@@ -580,62 +483,165 @@ def removeHandler(self, *args, **kwargs):
+    warn = warning
+
+    # pylint: disable=arguments-differ
+    def error(self, msg, exc_info=None, extra=None, sentry=False):   # type: ignore  # Signature of "debug" incompatible with supertype "LoggerAdapter"
```
and here

happz replied:
Fixed - explanation is now shared.

thrix commented on gluetool/log.py:
```diff
@@ -580,62 +483,165 @@ def removeHandler(self, *args, **kwargs):
+        self.log(logging.ERROR, msg, exc_info=exc_info, extra=extra, sentry=sentry)
+
+    # pylint: disable=arguments-differ
+    def exception(self, msg, exc_info=None, extra=None, sentry=False):   # type: ignore  # Signature of "debug" incompatible with supertype "LoggerAdapter"
```
and here ..

happz replied:
Fixed - explanation is now shared.

thrix commented on gluetool/glue.py:
```diff
@@ -470,18 +471,12 @@ class Pipeline(object):
     :ivar Module current_module: if set, it is the module which is currently being executed.
```
docs does not seem to reflect the reality?

thrix commented on gluetool/tool.py:
```diff
@@ -440,7 +440,7 @@ def run_pipeline(self):
 
             if failure and isinstance(failure.exception, GlueRetryError):
-                Glue.error(failure.exception)
+                Glue.error(str(failure.exception))
```
@happz why is this conversation needed now?

happz replied:
Because `Glue.error` accepts `str` as its first argument. An implicit conversion would happen, Python would call `str(failure.exception)` for us, but it wouldn't pass the type check since we were passing `Exception` where `str` is required. It wasn't a problem until now because `Glue.error` didn't have usable type annotation, that comes with this patch.

thrix reviewed (changes requested):
Some small changes requested, nothing serious

happz committed:
Share "type: ignore" explanation for logging methods

happz committed:
Move logger/methods setting into separate method

It is used by one of the depending libraries to re-set logging when more details about the context are known

optak reviewed (approved):
LGTM

thrix reviewed (approved):
LGTM, thanks for the fixes
